app/code/Magento/Swatches/Model/Plugin/EavAttribute.php

```diff
@@ -244,25 +253,44 @@ protected function saveSwatchParams(Attribute $attribute)
      * Save Visual Swatch data
      *
      * @param Attribute $attribute
+     *
      * @return void
      */
     protected function processVisualSwatch(Attribute $attribute)
     {
         $swatchArray = $attribute->getData('swatch/value');
+
         if (isset($swatchArray) && is_array($swatchArray)) {
-            foreach ($swatchArray as $optionId => $value) {
+            $visualSwatches = [];
+            foreach ($swatchArray as $key => $value) {
+                $visualSwatches[$key] =
+                    [
+                        self::DEFAULT_STORE_ID => $value,
+                        $this->storeManager->getStore()->getId() => '',
+                    ];
+            }
+
+            foreach ($visualSwatches as $optionId => $storeValues) {
                 $optionId = $this->getAttributeOptionId($optionId);
                 $isOptionForDelete = $this->isOptionForDelete($attribute, $optionId);
+
                 if ($optionId === null || $isOptionForDelete) {
                     //option was deleted by button with basket
                     continue;
                 }
-                $swatch = $this->loadSwatchIfExists($optionId, self::DEFAULT_STORE_ID);
 
-                $swatchType = $this->determineSwatchType($value);
+                $defaultSwatchValue = reset($storeValues);
+                foreach ($storeValues as $storeId => $value) {
+                    if (!$value) {
+                        $value = $defaultSwatchValue;
+                    }
 
-                $this->saveSwatchData($swatch, $optionId, self::DEFAULT_STORE_ID, $swatchType, $value);
-                $this->isSwatchExists = null;
+                    $swatch = $this->loadSwatchIfExists($optionId, $storeId);
+                    $swatchType = $this->determineSwatchType($value);
+
+                    $this->saveSwatchData($swatch, $optionId, $storeId, $swatchType, $value);
+                    $this->isSwatchExists = null;
```

Comment:
It is not really good idea for a Plugin to have a state like this.
Definitely not required in the scope of this Pull Request but whole class may be refactored, if you are interested in this.

Reply:
@ishakhsuvarov Unfortunately, I cannot find extra time for whole class refactoring for now. I agree with you this state is not a good idea, but if I remove it, it could affect other logic in this class. Is it ok to leave this until whole class will be refactored?